datafusion/expr/src/udf_docs.rs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/// Documentation for use by [`crate::ScalarUDFImpl`],
+/// [`crate::AggregateUDFImpl`] and [`crate::WindowUDFImpl`] functions
+/// that will be used to generate public documentation.
+///
+/// The name of the udf will be pulled from the [`crate::ScalarUDFImpl::name`],
+/// [`crate::AggregateUDFImpl::name`] or [`crate::WindowUDFImpl::name`] function
+/// as appropriate.
+///
+/// All strings in the documentation are required to be
+/// in [markdown format](https://www.markdownguide.org/basic-syntax/).
+///
+/// Currently, documentation only supports a single language
+/// thus all text should be in English.
+#[derive(Debug, Clone)]
```

Comment:
What do you think about changing `Documentation` and `DocSection` from requiring `&'static str` to `String`? 

It would be slightly less convenient to make singleton documentation (we would have to use `OnceLock` or something similar) but would allow dynamically created documentation content? 

I think it would also allow some nicer builder style APIs (like adding arguments with multiple function calls)

Reply:
Heh, so originally I had it that way but thought that since docs were almost always going to be static and String can't be used in const (but can in static with OnceLock or soon(ish) lazylock) this approach might be the best. I literally rewote it 3 times trying to find a perfect solution ... and now I'm not sure I found it :) What made me decide to go with the static approach was that I encountered a situation where I would have to clone docs in the `documentation(&self) ->&Documentation` function which didn't feel right to me. I'm all for getting this right first before merging this PR though ... and I really do like the builder idea.

Reply:
I updated Documentation but I left DocSection alone for now.